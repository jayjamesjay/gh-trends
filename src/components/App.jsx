import React, { Component } from 'react';
import { BrowserRouter as Router, Route } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';

import Footer from './Footer';
import Header from './Header';
import Nav from './Nav';

import Home from '../routes/Home';
import Search from '../routes/Search';
import Saved from '../routes/Saved';

import GlobalStyle from '../styles/Global';
import { Main } from '../styles/Main';
import { dark, light } from '../styles/Theme';

const links = [['Search', '/search'], ['Saved', '/saved']];

export default class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      hideMenu: true,
      theme: light,
      saved: []
    };
  }

  toggleMenu = () => {
    const { hideMenu } = this.state;
    this.setState({
      hideMenu: !hideMenu
    });
  };

  switchTheme = () => {
    const { theme } = this.state;
    this.setState({
      theme: theme === light ? dark : light
    });
  };

  save = elem => {
    const { saved } = this.state;
    let currSaved = saved.slice();

    if (currSaved.findIndex(item => item.nameWithOwner === elem.nameWithOwner) < 0) {
      currSaved.push(elem);
    } else {
      currSaved = currSaved.filter(item => item.nameWithOwner !== elem.nameWithOwner);
    }

    this.setState({
      saved: currSaved
    });
  };

  removeAll = () => {
    this.setState({
      saved: []
    });
  };

  render() {
    const {
      state: { hideMenu, theme, saved },
      toggleMenu,
      switchTheme,
      save,
      removeAll
    } = this;

    return (
      <Router>
        <ThemeProvider theme={theme}>
          <>
            <Header
              link="/"
              title="GH Trends"
              toggle={toggleMenu}
              hide={hideMenu}
              switchTheme={switchTheme}
            />
            <Nav links={links} hide={hideMenu} linkClick={toggleMenu} />
            <Main>
              <Route path="/search" render={() => <Search save={save} saved={saved} />} />
              <Route
                path="/saved"
                render={() => <Saved data={saved} save={save} removeAll={removeAll} />}
              />
              <Route exact path="/" render={() => <Home save={save} saved={saved} />} />
              <Route path="/home" render={() => <Home save={save} saved={saved} />} />
            </Main>
            <Footer />
            <GlobalStyle />
          </>
        </ThemeProvider>
      </Router>
    );
  }
}
