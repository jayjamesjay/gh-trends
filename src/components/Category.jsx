import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { LabelSpan } from '../styles/Span';
import { InputRadio } from '../styles/Input';
import LabelCategory from '../styles/Label';

/**
 * @module Category
 */

/**
 * Single category selector
 *
 * @property {string} label - name of the category
 * @property {boolean} checked - determinates if category is active
 * @property {function} onClick - function fired when category is clicked
 * @returns {Category}
 */
export default function Category({ label, checked, onClick }) {
  const click = useCallback(() => onClick(label), [label, onClick]);

  return (
    <div onClick={click} role="radio" aria-checked={checked} onKeyPress={click} tabIndex="0">
      <LabelCategory active={checked} htmlFor={label}>
        <LabelSpan>{label}</LabelSpan>
      </LabelCategory>
      <InputRadio
        id={label}
        type="radio"
        value={label}
        checked={checked}
        onChange={click}
        tabIndex="-1"
      />
    </div>
  );
}

Category.propTypes = {
  label: PropTypes.string.isRequired,
  checked: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired
};
