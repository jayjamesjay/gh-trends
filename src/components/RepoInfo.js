/**
 * @module RepoInfo
 */

/**
 * Information about repository
 * @class
 */
class RepoInfo {
  /**
   * Creates a RepoInfo
   *
   * @param {string} nameWithOwner - name and author of repository
   * @param {string} url - url of the repository
   * @param {string} description - description
   * @param {number} stargazersCount - amount of stars
   * @param {string} language - programming language
   * @param {number} forks - amount of forks
   * @param {string} license - repository license
   */
  constructor(nameWithOwner, url, description, stargazersCount, language, forks, license) {
    this.nameWithOwner = nameWithOwner;
    this.url = url;
    this.description = description;
    this.stargazersCount = stargazersCount;
    this.language = language;
    this.forks = forks;
    this.license = license;
  }

  /**
   * Creates RepoInfo from Github API data
   *
   * @param {object} item - repository data from Github API
   * @returns {RepoInfo}
   */
  static fromGithubRes(item) {
    const { description, language, forks } = item;
    const license = item.license ? item.license.spdx_id : null;

    return new RepoInfo(
      item.full_name,
      item.html_url,
      description,
      item.stargazers_count,
      language,
      forks,
      license
    );
  }
}

export default RepoInfo;
