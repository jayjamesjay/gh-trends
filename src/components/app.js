import React, { Component } from 'react';
import { BrowserRouter as Router, Route } from 'react-router-dom';

import Footer from './footer';
import Header from './header';
import Nav from './nav';

import Home from '../routes/home';
import Search from '../routes/search';
import Saved from '../routes/saved';

import { Main, GlobalStyle } from './styles';
import { dark, light } from './theme';
import { ThemeProvider } from 'styled-components';

const links = [['Search', '/search'], ['Saved', '/saved']];

export default class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      hideMenu: true,
      theme: light,
    };
  }

  toggleMenu = () => {
    this.setState({
      hideMenu: !this.state.hideMenu,
    });
  };

  switchTheme = () => {
    this.setState({
      theme: this.state.theme === light ? dark : light,
    });
  };

  render() {
    return (
      <Router>
        <ThemeProvider theme={this.state.theme}>
          <>
            <Header
              link="/"
              title="GH Trends"
              toggle={this.toggleMenu}
              hide={this.state.hideMenu}
              switchTheme={this.switchTheme}
            />
            <Nav
              links={links}
              hide={this.state.hideMenu}
              linkClick={this.toggleMenu}
            />
            <Main>
              <Route path="/search" component={Search} />
              <Route path="/saved" component={Saved} />
              <Route exact path="/" component={Home} />
              <Route path="/home" component={Home} />
            </Main>
            <Footer />
            <GlobalStyle />
          </>
        </ThemeProvider>
      </Router>
    );
  }
}
