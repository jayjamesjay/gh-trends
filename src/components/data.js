import formatDate from "../components/date";

const weekAgo = new Date();
weekAgo.setDate(weekAgo.getDate() - 8);
const queryWeek = `sort:stars-desc created:>${formatDate(weekAgo)}`;

const monthAgo = new Date();
monthAgo.setMonth(monthAgo.getMonth() - 1);
const queryMonth = `sort:stars-desc created:>${formatDate(monthAgo)}`;

const queryAllTime = "stars:>1000 sort:stars-desc";
export const queryList = [queryWeek, queryMonth, queryAllTime];

const fill = {
  nameWithOwner: "jayjamesjay/gh-trends",
  url: "",
  description: "Loading content for this website...",
  stargazers: {
    totalCount: 123
  },
  repositoryTopics: {
    nodes: [
      {
        topic: {
          name: "Just"
        }
      },
      {
        topic: {
          name: "a"
        }
      },
      {
        topic: {
          name: "moment"
        }
      }
    ]
  },
  languages: {
    nodes: []
  }
};

export const initData = new Array(9);
initData.fill(fill, 0, 9);

export default class Data {
  constructor(id, data, cursor) {
    this.id = id;
    this.data = data;
    this.cursor = cursor;
  }
}