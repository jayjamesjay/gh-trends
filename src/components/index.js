import React, { Component } from "react";
import { render } from "react-dom";
import App from "./app";

if ("serviceWorker" in navigator) {
  navigator.serviceWorker.register("worker.js");
}

render(<App />, document.getElementById("app"));
