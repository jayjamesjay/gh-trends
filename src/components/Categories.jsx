import React from 'react';
import PropTypes from 'prop-types';

import { CategoryMenu } from '../styles/Form';
import Category from './Category';

/**
 * @module Categories
 */

/**
 * Menu to hadle category selection
 *
 * @property {array} labels - names of categories
 * @property {string} active - name of label which is currently active
 * @property {function} onClick - function called whenever label is clicked
 * @returns {Categories}
 */
export default function Categories({ labels, active, onClick }) {
  return (
    <CategoryMenu>
      {labels.map((label) => (
        <Category checked={active === label} key={label} label={label} onClick={onClick} />
      ))}
    </CategoryMenu>
  );
}

Categories.propTypes = {
  labels: PropTypes.arrayOf(PropTypes.string).isRequired,
  active: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
};
