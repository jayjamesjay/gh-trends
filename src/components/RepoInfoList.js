import RepoInfo from './RepoInfo';
/**
 * @module RepoInfoList
 */

/**
 * List of information about repositories
 * @class
 */
class RepoInfoList {
  /**
   * Creates a RepoInfoList
   *
   * @param {array} data - main content
   * @param {number} page - last loaded page
   */
  constructor(data, page) {
    this.data = data;
    this.page = page;
  }

  /**
   * Creates data part of RepoInfoList from Github API data
   *
   * @param {array} data - list of repository data from Github API
   * @returns {array}
   */
  static fromGithubRes(data) {
    return data.map((item) => RepoInfo.fromGithubRes(item));
  }

  /**
   * Parses and adds Github API data to this RepoInfoList
   * and increases the page by 1.
   *
   * @param {array} data - list of repository data from Github API
   */
  update(data) {
    const content = RepoInfoList.fromGithubRes(data);
    this.data = this.data.concat(content);
    this.page += 1;
  }
}

export default RepoInfoList;
