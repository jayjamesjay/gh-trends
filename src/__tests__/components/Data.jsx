import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import { jsonToMarkdown, DownloadLink, identicalItems } from '../../components/Data';
import RepoInfoList from '../../utils/RepoInfoList';
import RepoInfo from '../../utils/RepoInfo';

it('compares nameWithOwner of two items', () => {
  const elem1 = { nameWithOwner: 'abb', age: 30 };
  const elem2 = { nameWithOwner: 'abb', age: 31 };

  expect(identicalItems(elem1, elem2)).toBe(true);
});

describe('RepoInfo', () => {
  it('from one element of Github Reest API reponse array', () => {
    const response = {
      full_name: 'jayjamesjay/gh-trends',
      html_url: '',
      description: 'Loading content for this website...',
      stargazers_count: 100,
      language: 'JavaScript',
      forks: 321,
      license: {
        spdx_id: 'MIT',
      },
    };

    const expected = new RepoInfo(
      'jayjamesjay/gh-trends',
      '',
      'Loading content for this website...',
      100,
      'JavaScript',
      321,
      'MIT'
    );

    expect(RepoInfo.fromGithubRes(response)).toEqual(expected);
  });

  it('from one element of Github Reest API reponse array without license', () => {
    const response = {
      full_name: 'jayjamesjay/gh-trends',
      html_url: '',
      description: 'Loading content for this website...',
      stargazers_count: 100,
      language: 'JavaScript',
      forks: 321,
      license: null,
    };

    const expected = new RepoInfo(
      'jayjamesjay/gh-trends',
      '',
      'Loading content for this website...',
      100,
      'JavaScript',
      321,
      null
    );

    expect(RepoInfo.fromGithubRes(response)).toEqual(expected);
  });
});

describe('<RepoInfoList />', () => {
  const expectedElem = new RepoInfo(
    'jayjamesjay/gh-trends',
    '',
    'Loading content for this website...',
    100,
    'JavaScript',
    321,
    'MIT'
  );

  const expectedData = new Array(3);
  expectedData.fill(expectedElem);

  it('from Github Reest API reponse', () => {
    const respElem = {
      full_name: 'jayjamesjay/gh-trends',
      html_url: '',
      description: 'Loading content for this website...',
      stargazers_count: 100,
      language: 'JavaScript',
      forks: 321,
      license: {
        spdx_id: 'MIT',
      },
    };

    const response = new Array(3);
    response.fill(respElem);

    expect(RepoInfoList.fromGithubRes(response)).toEqual(expectedData);
  });

  it('upates current data', () => {
    const respElem = {
      full_name: 'jayjamesjay/gh-trends',
      html_url: '',
      description: 'Loading content for this website...',
      stargazers_count: 100,
      language: 'JavaScript',
      forks: 321,
      license: {
        spdx_id: 'MIT',
      },
    };
    const response = [respElem];
    const infoList = new RepoInfoList(expectedData, 1);
    const expectedInfoList = new RepoInfoList(expectedData.concat(expectedElem), 2);
    infoList.update(response);

    expect(infoList).toEqual(expectedInfoList);
  });

  it('constructs full with id', () => {
    const page = 1;
    const infoList = new RepoInfoList(expectedData, page);
    const expectedList = {
      data: expectedData,
      page,
    };

    expect(infoList).toEqual(expectedList);
  });
});

describe('<DownloadLink />', () => {
  const name = 'msg.txt';
  const text = `You've just downloaded this text ;)`;
  const title = 'Download message';

  it(`renders with default style`, () => {
    const component = renderer.create(
      <DownloadLink filename={name} content={text} display={title} />
    );
    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('displays title', () => {
    render(<DownloadLink filename={name} content={text} display={title} />);
    expect(screen.getByRole('link').textContent).toEqual(title);
  });

  it('points to content', () => {
    const data = `data:plain/plain;charset=utf-8,${encodeURIComponent(text)}`;
    render(<DownloadLink filename={name} content={text} display={title} />);

    expect(screen.getByRole('link').href).toEqual(data);
  });
});

it('converts JSON to Markdown', () => {
  const json = [
    { name: 'Dog', length: 1 },
    { name: 'Sharp', length: 22 },
    { name: 'Cat', length: 2 },
  ];

  expect(jsonToMarkdown(json)).toBe(
    '# Dog \r\n' +
      'length: 1 \r\n' +
      '\r\n' +
      '\r\n' +
      '# Sharp \r\n' +
      'length: 22 \r\n' +
      '\r\n' +
      '\r\n' +
      '# Cat \r\n' +
      'length: 2 \r\n' +
      '\r\n' +
      '\r\n'
  );
});
