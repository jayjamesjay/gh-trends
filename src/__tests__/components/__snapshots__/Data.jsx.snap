// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DownloadLink /> renders with default style 1`] = `
.c0 {
  display: block;
  font-size: 1.2rem;
  font-family: 'Noto Sans HK',sans-serif;
  color: #000;
  text-decoration: none;
  cursor: pointer;
}

<a
  className="c0"
  download="msg.txt"
  href="data:plain/plain;charset=utf-8,You've%20just%20downloaded%20this%20text%20%3B)"
>
  Download message
</a>
`;
